modem: backend: uart: Add hw-flow-control for UART

MarkusLassila opened the pull request:
Add Asynchronous UART implementation, which does not drop data when automatic hardware-flow-control is set in the device tree.

With automatic hardware flow control, the CTS pin will be automatically deactivated when there are no more asynchronous UART RX buffers available. After buffer space becomes available, and UART RX is restarted, the CTS pin will be activated.

MarkusLassila committed:
modem: backend: uart: Add hw-flow-control for UART

Add Asynchronous UART implementation, which does not drop data
when automatic hardware-flow-control is set in the device tree.

With automatic hardware flow control, the CTS pin will be
automatically deactivated when there are no more asynchronous
UART RX buffers available. After buffer space becomes available,
and UART RX is restarted, the CTS pin will be activated.

Signed-off-by: Markus Lassila <markus.lassila@nordicsemi.no>

bjarki-andreasen reviewed (changes requested):
Could you try making the receive buffers smaller in the uart backend test suite https://github.com/zephyrproject-rtos/zephyr/blob/044a94f06a0fa5f1394751dafca8ae1a8cc5004b/tests/subsys/modem/backends/uart/src/main.c#L39 and add a sleep here https://github.com/zephyrproject-rtos/zephyr/blob/044a94f06a0fa5f1394751dafca8ae1a8cc5004b/tests/subsys/modem/backends/uart/src/main.c#L137 to try to invoke the need for flowcontrol?


 I have a suspicion bytes may be lost by disabling RX outright rather than using flow control (as disable may flush the rx buffer) :)

SeppoTakalo commented:
I tried this a little bit.

Using nRF9151, I'm now able to reliably use 1 Mbit/s baud rate on the serial port without losing data on serial buffers.
So this kind of implementation significantly improves the upload performance.

bjarki-andreasen-nordic committed:
tests: modem: backends: uart: boards: add hw-flow-control

Update the board overlays to support hardware flow control for
testing.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: modem: backends: uart: add hw flow control option

Extend test modem uart backend test suite to support testing hw
flow control, which is performed by using a small receive
buffer for the modem uart backend, and slowing down the read
of received data, ensuring the buffer will be overrun if hw flow
control is not working.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen reviewed (changes requested):
Please cherry pick updated test from https://github.com/zephyrproject-rtos/zephyr/pull/89250 and add a test case which enables the new CONFIG_TEST_HW_FLOW_CONTROL option.

MarkusLassila commented on tests/subsys/modem/backends/uart/testcase.yaml:
```diff
@@ -14,6 +14,12 @@ tests:
     extra_configs:
       - CONFIG_UART_ASYNC_API=y
 
+  modem.backends.uart.async.hwfc:
```
@bjarki-andreasen: The tes case uses gpio_loopback fixture, so I am not sure how it can ever pass in the CI as there are not suitable DK's which would have the CTS and RTS wired together?

bjarki-andreasen replied:
The nrf5340_cpuapp gpio_loopback fixture in downstream CI has P0.4 + P0.5 and P0.6 + P0.7 so it should work perfectly :)

MarkusLassila replied:
Ah, so you have a control over the nRF5340DK that serves for the tests. Out of curiosity, is there only singular DK wired for these tests?

MarkusLassila replied:
Discussed with Bjarki, the DK's in nordic CI have the loopback wires in place.

MarkusLassila commented:
@bjarki-andreasen: For whatever the reason, it does not seem that the CI ran the modem.backends tests.

tomi-font reviewed (changes requested):
Reviewed all but `modem_backend_uart_async_hwfc.c`. Looks good overall, just some minor remarks.

MarkusLassila commented on subsys/modem/backends/Kconfig:
```diff
@@ -48,6 +48,29 @@ config MODEM_BACKEND_UART_ASYNC_RECEIVE_IDLE_TIMEOUT_MS
+config MODEM_BACKEND_UART_ASYNC_HWFC_DISABLED
+	bool "Hardware flow control is disabled"
+
+config MODEM_BACKEND_UART_ASYNC_HWFC_ENABLED
```
The UART backend works fine without hw-flow-control property as well. I don't think that we can check out that the particular UART which would be used in here would have the hw-flow-control as the selection of the UART is left for the application.

tomi-font replied:
Even though it's the application's choice I think you might be able to get it by using DT chosen properties and the like... But we can do without. Defaulting to the non-HWFC backend is enough.

MarkusLassila replied:
Let's leave this as a future exercise then.

MarkusLassila commented on subsys/modem/backends/Kconfig:
```diff
@@ -48,6 +48,29 @@ config MODEM_BACKEND_UART_ASYNC_RECEIVE_IDLE_TIMEOUT_MS
 	int "Modem async UART receive idle timeout in milliseconds"
 	default 30
 
+choice MODEM_BACKEND_UART_ASYNC_HWFC_CHOICE
+	prompt "Modem async UART hardware flow control"
+	default MODEM_BACKEND_UART_ASYNC_HWFC_DISABLED
+	help
+	  Select whether to enable or disable hardware flow control (HWFC)
+	  for the modem async UART backend.
+
+config MODEM_BACKEND_UART_ASYNC_HWFC_DISABLED
+	bool "Hardware flow control is disabled"
+
+config MODEM_BACKEND_UART_ASYNC_HWFC_ENABLED
+	bool "Hardware flow control is enabled"
+	select EXPERIMENTAL
+endchoice
```
That is for CMake. We have 3 options, ISR, ASYNC and ASYNC_HWFC, so I think that I would need to use the if(), elseif(), else() for the Kconfigs.

I can do that if you think that it is cleaner to do that way.

tomi-font replied:
Ah I see now. Yeah, even though you'll need to use ifs in CMake, I think it's better to do it that way. It will keep the configuration more understandable. Choices are meant for choosing between options, and even though you could argue enabled or disabled are two options, really a single Kconfig option is most fitted to that.

MarkusLassila commented on tests/subsys/modem/backends/uart/Kconfig:
```diff
@@ -0,0 +1,16 @@
+
+source "Kconfig.zephyr"
+
+config TEST_HW_FLOW_CONTROL
```
@bjarki-andreasen?

MarkusLassila commented on tests/subsys/modem/backends/uart/testcase.yaml:
```diff
@@ -5,15 +5,26 @@ common:
   modem.backends.uart.async:
     extra_configs:
       - CONFIG_UART_ASYNC_API=y
+    platform_allow:
+      - b_u585i_iot02a
+      - nrf5340dk/nrf5340/cpuapp
+
+  modem.backends.uart.async.hwfc:
+    extra_configs:
+      - CONFIG_UART_ASYNC_API=y
+      - CONFIG_MODEM_BACKEND_UART_ASYNC_HWFC_ENABLED=y
+      - CONFIG_TEST_HW_FLOW_CONTROL=y
+    platform_allow:
+      - nrf5340dk/nrf5340/cpuapp
 
   modem.backends.uart.isr:
     extra_configs:
       - CONFIG_UART_INTERRUPT_DRIVEN=y
+    platform_allow:
```
The reason is not timing or resource constraint in this case. The UART driver for STM32 does not support hw-flow-control so that it would not lose data.

MarkusLassila replied:
I think that the question is whether there should be platform_allow on the common level and the integration_platform would then be applied in the test level. Given that there is no resource constraint, I would like to keep this as it is.

tomi-font commented on include/zephyr/modem/backend/uart.h:
```diff
@@ -30,17 +30,37 @@ struct modem_backend_uart_isr {
 	atomic_t state;
 };
 
+#ifdef CONFIG_MODEM_BACKEND_UART_ASYNC_HWFC
+struct rx_queue_event {
+	uint8_t *buf;
+	size_t len;
+};
+
+struct modem_backend_uart_async {
+	struct modem_backend_uart_async_common common;
+	struct k_mem_slab rx_slab;
+	struct k_msgq rx_queue;
+	struct rx_queue_event rx_event;
+	struct rx_queue_event rx_queue_buf[CONFIG_MODEM_BACKEND_UART_ASYNC_HWFC_BUFFER_COUNT];
+	uint32_t rx_buf_size;
+	uint8_t rx_buf_count;
+};
+#else
+struct modem_backend_uart_async {
+	struct modem_backend_uart_async_common common;
+	uint8_t *receive_bufs[2];
+	uint32_t receive_buf_size;
+	struct ring_buf receive_rb;
+	struct k_spinlock receive_rb_lock;
+};
+#endif /* CONFIG_MODEM_BACKEND_UART_ASYNC_HWFC */
```
Could have some spacing around those `#ifdef`/`#else`/`#endif`.

MarkusLassila commented:
@tomi-font, @SeppoTakalo: Seems that a minor change dismissed your reviews. Apologies.
